File: backend/api/motor.py
"""HTTP routes for the trombone motor. All logic lives in `backend.motor`."""

from __future__ import annotations

import asyncio

from fastapi import APIRouter
from pydantic import BaseModel, Field

from .. import motor as svc
from .errors import handle_driver_errors

router = APIRouter(prefix="/motor", tags=["motor"])


class MotorStatus(BaseModel):
    connected: bool
    moving: bool = False
    position: int | None = None
    device_index: int | None = None
    soft_min: int | None = None
    soft_max: int | None = None
    error: str | None = None


class LimitsRequest(BaseModel):
    soft_min: int | None = Field(default=None, description="Min travel limit (pulses); null = unbounded")
    soft_max: int | None = Field(default=None, description="Max travel limit (pulses); null = unbounded")


class ConnectRequest(BaseModel):
    device_index: int = Field(default=0, ge=0, le=15)


class MoveRequest(BaseModel):
    position: int = Field(..., description="Target position in pulses")
    absolute: bool = Field(default=True)


class HomeRequest(BaseModel):
    direction: str = Field(..., pattern="^(positive|negative)$")


class JogRequest(BaseModel):
    positive: bool = Field(..., description="True = jog toward max, False = toward min")
    speed: int | None = Field(default=None, ge=1, le=5000, description="Jog speed (pulses/s)")


class DiscoverResponse(BaseModel):
    candidates: list[str]


async def _status() -> MotorStatus:
    with handle_driver_errors("motor status"):
        return MotorStatus(**await asyncio.to_thread(svc.read_status))


@router.get("/status", response_model=MotorStatus)
async def status() -> MotorStatus:
    return await _status()


@router.get("/discover", response_model=DiscoverResponse)
async def discover() -> DiscoverResponse:
    with handle_driver_errors("motor discover"):
        cands = await asyncio.to_thread(svc.discover)
    return DiscoverResponse(candidates=cands)


@router.post("/connect", response_model=MotorStatus)
async def connect(req: ConnectRequest) -> MotorStatus:
    with handle_driver_errors("motor connect"):
        await asyncio.to_thread(svc.connect, req.device_index)
    return await _status()


@router.post("/disconnect", response_model=MotorStatus)
async def disconnect() -> MotorStatus:
    with handle_driver_errors("motor disconnect"):
        await asyncio.to_thread(svc.disconnect)
    return MotorStatus(connected=False)


@router.post("/move", response_model=MotorStatus)
async def move(req: MoveRequest) -> MotorStatus:
    with handle_driver_errors("motor move"):
        await asyncio.to_thread(svc.move, req.position, req.absolute)
    return await _status()


@router.post("/home", response_model=MotorStatus)
async def home(req: HomeRequest) -> MotorStatus:
    with handle_driver_errors("motor home"):
        await asyncio.to_thread(svc.home, req.direction == "positive")
    return await _status()


@router.post("/limits", response_model=MotorStatus)
async def set_limits(req: LimitsRequest) -> MotorStatus:
    with handle_driver_errors("motor set limits"):
        await asyncio.to_thread(svc.set_limits, req.soft_min, req.soft_max)
    return await _status()


@router.post("/jog/start", response_model=MotorStatus)
async def jog_start(req: JogRequest) -> MotorStatus:
    with handle_driver_errors("motor jog start"):
        await asyncio.to_thread(svc.jog_start, req.positive, req.speed)
    return await _status()


@router.post("/jog/stop", response_model=MotorStatus)
async def jog_stop() -> MotorStatus:
    with handle_driver_errors("motor jog stop"):
        await asyncio.to_thread(svc.jog_stop)
    return await _status()


@router.post("/stop", response_model=MotorStatus)
async def stop() -> MotorStatus:
    with handle_driver_errors("motor stop"):
        await asyncio.to_thread(svc.stop)
    return await _status()

File: frontend/src/api/motor.ts
import { http } from './client'

export interface MotorStatus {
  connected: boolean
  moving: boolean
  position: number | null
  device_index: number | null
  error: string | null
  soft_min: number | null
  soft_max: number | null
}

export const motor = {
  status: () => http<MotorStatus>('/motor/status'),
  discover: () => http<{ candidates: string[] }>('/motor/discover'),
  connect: (deviceIndex = 0) =>
    http<MotorStatus>('/motor/connect', {
      method: 'POST',
      body: JSON.stringify({ device_index: deviceIndex }),
    }),
  disconnect: () =>
    http<MotorStatus>('/motor/disconnect', { method: 'POST' }),
  move: (position: number, absolute = true) =>
    http<MotorStatus>('/motor/move', {
      method: 'POST',
      body: JSON.stringify({ position, absolute }),
    }),
  home: (direction: 'positive' | 'negative') =>
    http<MotorStatus>('/motor/home', {
      method: 'POST',
      body: JSON.stringify({ direction }),
    }),
  jogStart: (positive: boolean, speed?: number) =>
    http<MotorStatus>('/motor/jog/start', {
      method: 'POST',
      body: JSON.stringify({ positive, speed: speed ?? null }),
    }),
  jogStop: () => http<MotorStatus>('/motor/jog/stop', { method: 'POST' }),
  setLimits: (softMin: number | null, softMax: number | null) =>
    http<MotorStatus>('/motor/limits', {
      method: 'POST',
      body: JSON.stringify({ soft_min: softMin, soft_max: softMax }),
    }),
  stop: () => http<MotorStatus>('/motor/stop', { method: 'POST' }),
}
